### PR TITLE
Autogenerated docker-compose.yml

### PR DESCRIPTION
**Pull request by lotas**
This makes generation of `docker-compose.yml` part of the `yarn generate`.

`docker-compose up -d` allows to run everything locally (without background jobs for now)

Taskcluster is exposed through nginx ingress service at `http://localhost:8080/`.

Login is possible using hardcoded `static/taskcluster/*` credentials that can be seen in (`docker-compose.yml`)

TODO:
- [x] generate docker-compose with services
- [x] ui
- [x] auth-web
- [x] github-web (github credentials missing)
- [x] hooks-web
- [x] index-web
- [x] notify-web
- [x] object-web
- [x] purge-cache-web
- [x] queue-web (bucket missing)
- [x] secrets-web
- [x] web-server-web
- [x] worker-manager-web
- [x] single ingress
- [x] built-in workers

Background tasks are currently not included.
They can be run manually when needed using:

```sh
docker-compose run --rm auth-web worker-manager/expire-workers
# or
docker-compose exec auth-web worker-manager/expire-workers
```

### CODE REVIEW

**Comment by Eijebong** [comment]
You probably want to add a healthcheck for postgres to make sure it's completely started before services depending on it.

```diff
diff --git a/docker-compose.yml.sample b/docker-compose.yml.sample
index fe734978f..1144fccfd 100644
--- a/docker-compose.yml.sample
+++ b/docker-compose.yml.sample
@@ -22,6 +22,12 @@ services:
       - local
     ports:
       - '5432:5432'
+    healthcheck:
+      test: pg_isready
+      interval: 1s
+      timeout: 2s
+      retries: 20
+      start_period: 3s
     environment:
       POSTGRES_HOST_AUTH_METHOD: trust
       LC_COLLATE: en_US.UTF8
```

**Comment by Eijebong** [comment]
Diff that includes postgres migrations:
```diff
diff --git a/docker-compose.yml.sample b/docker-compose.yml.sample
index fe734978f..fba39f971 100644
--- a/docker-compose.yml.sample
+++ b/docker-compose.yml.sample
@@ -14,6 +14,12 @@ services:
     ports:
       - '5672:5672'
       - '15672:15672'
+    healthcheck:
+      test: rabbitmq-diagnostics ping
+      interval: 1s
+      timeout: 2s
+      retries: 20
+      start_period: 3s
   postgres:
     image: postgres:11
     env_file:
@@ -22,10 +28,58 @@ services:
       - local
     ports:
       - '5432:5432'
+    healthcheck:
+      test: pg_isready -U postgres
+      interval: 1s
+      timeout: 2s
+      retries: 20
+      start_period: 3s
     environment:
       POSTGRES_HOST_AUTH_METHOD: trust
       LC_COLLATE: en_US.UTF8
       LC_CTYPE: en_US.UTF8
+  pg_migrate_1:
+    image: postgres:11
+    env_file:
+      - .env
+    networks:
+      - local
+    depends_on:
+      postgres:
+        condition: service_healthy
+    command:
+      - /bin/bash
+      - -c
+      - "psql -h postgres -U postgres -c \"CREATE DATABASE test_taskcluster\";
+         if [[ $$? == 0 ]]; then
+           psql -h postgres -U postgres -c \";
+           CREATE USER test PASSWORD 'test';
+           GRANT ALL ON DATABASE test_taskcluster TO test WITH GRANT OPTION;
+           CREATE USER test_auth PASSWORD 'auth_password';
+           CREATE USER test_github PASSWORD 'github_password';
+           CREATE USER test_hooks PASSWORD 'hooks_password';
+           CREATE USER test_index PASSWORD 'index_password';
+           CREATE USER test_notify PASSWORD 'notify_password';
+           CREATE USER test_object PASSWORD 'object_password';
+           CREATE USER test_purge_cache PASSWORD 'purge_cache_password';
+           CREATE USER test_queue PASSWORD 'queue_password';
+           CREATE USER test_secrets PASSWORD 'secrets_password';
+           CREATE USER test_web_server PASSWORD 'web_server_password';
+           CREATE USER test_worker_manager PASSWORD 'worker_manager_password';
+           \";
+         fi;
+         exit 0"
+  pg_migrate_2:
+    image: taskcluster/taskcluster:v44.13.5
+    env_file:
+      - .env
+    networks:
+      - local
+    depends_on:
+      pg_migrate_1:
+        condition: service_completed_successfully
+    command:
+      - script/db:upgrade
   ui:
     image: taskcluster/taskcluster:v44.13.5
     env_file:
@@ -41,8 +95,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - auth/web
   github-web:
@@ -52,8 +108,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - github/web
   hooks-web:
@@ -63,8 +121,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - hooks/web
   index-web:
@@ -74,8 +134,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - index/web
   notify-web:
@@ -85,8 +147,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - notify/web
   object-web:
@@ -96,8 +160,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - object/web
   purge-cache-web:
@@ -107,8 +173,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - purge-cache/web
   queue-web:
@@ -118,8 +186,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - queue/web
   secrets-web:
@@ -129,8 +199,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - secrets/web
   web-server-web:
@@ -140,8 +212,10 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     ports:
       - '3050:3050'
     command:
@@ -153,7 +227,9 @@ services:
     networks:
       - local
     depends_on:
-      - rabbitmq
-      - postgres
+      rabbitmq:
+        condition: service_healthy
+      pg_migrate_2:
+        condition: service_completed_successfully
     command:
       - worker-manager/web

```

It seems to be slightly angry though:
```
taskcluster-postgres-1            | 2022-05-07 14:06:32.923 UTC [93] ERROR:  function "online_migration_v2_batch" does not exist at character 8
taskcluster-postgres-1            | 2022-05-07 14:06:32.923 UTC [93] STATEMENT:  select 'online_migration_v2_batch'::regproc
```

**Comment by Eijebong** [comment]
Also need to setup rabbimtq user/pass/vhost:

```
environment:
  RABBITMQ_DEFAULT_USER: admin
  RABBITMQ_DEFAULT_PASS: admin
  RABBITMQ_DEFAULT_VHOST: local
```

**Comment by Eijebong** [comment]
The `ui` container is missing its command.

```
command:
  - ui/web
```

**Comment by Eijebong** [comment]
Pushed my progress here. https://github.com/Eijebong/taskcluster/blob/docker-compose-progress/docker-compose.yml

**Comment by Eijebong** [comment]
Adding this to `.env` applies it to all services, leading to new errors :)

```
FORCE_SSL=false
TRUST_PROXY=true
```

**Comment by lotas** [comment]
Awesome @Eijebong, thanks for the ideas and suggestions. I'll have a look.
True, I didn't finish all the env variables yet, so it is not working at the moment.

**Comment by Eijebong** [comment]
With my progress I was able to log in. I got stuck at getting graphql to answer properly but then I haven't searched for long

**Comment by lotas** [comment]
> With my progress I was able to log in. I got stuck at getting graphql to answer properly but then I haven't searched for long

I will borrow your ideas @Eijebong and apply them here, thanks!
Still WIP, working on other features in parallel

**Comment by lotas** [comment]
> For minio to work properly:
> 
> #5042 #5043

oh great! Thanks for linking those.

**Comment by lotas** [comment]
```
image: quay.io/minio/minio
environment:
  - MINIO_ROOT_USER=S3USER
  - MINIO_ROOT_PASSWORD=S3PASSWORD
command: "server /data --console-address ':9001'"
volumes:
  - ./s3data:/data
ports:
  - "9000:9000"
  - "9001:9001"
  - ```

**Comment by Eijebong** [comment]
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 7efce1f6d..5c446d181 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -68,6 +68,12 @@ services:
       - '3091:9001'
     volumes:
       - ./docker/buckets:/data
+    healthcheck:
+      test: curl -I http://s3:9000/minio/health/cluster
+      interval: 1s
+      timeout: 2s
+      retries: 30
+      start_period: 3s
   s3_init_buckets:
     image: minio/mc
     networks:
@@ -76,10 +82,20 @@ services:
       MINIO_ENDPOINT: http://s3:9000
       MINIO_ROOT_USER: minioadmin
       MINIO_ROOT_PASSWORD: miniopassword
+    entrypoint: >
+      /bin/sh -c "
+      /usr/bin/mc config host rm local;
+      /usr/bin/mc config host add --quiet --api s3v4 local http://s3:9000 minioadmin miniopassword;
+      /usr/bin/mc rb --force local/public-bucket/;
+      /usr/bin/mc mb --quiet local/public-bucket/;
+      /usr/bin/mc policy set public local/public-bucket;
+      /usr/bin/mc rb --force local/private-bucket/;
+      /usr/bin/mc mb --quiet local/private-bucket/;
+      /usr/bin/mc policy set public local/public-bucket;
+      "
     depends_on:
-      - s3
-    command:
-      - mb
+      s3:
+        condition: service_healthy
   ui:
     image: taskcluster/taskcluster:v44.13.5
     networks:
@@ -244,8 +260,11 @@ services:
       NODE_ENV: development
       FORCE_SSL: 'false'
       TRUST_PROXY: 'true'
-      AWS_ACCESS_KEY_ID: ''
-      AWS_SECRET_ACCESS_KEY: ''
+      AWS_ACCESS_KEY_ID: 'minioadmin'
+      AWS_SECRET_ACCESS_KEY: 'miniopassword'
+      AWS_FORCE_PATH_STYLE: 'true'
+      AWS_SKIP_CORS_CONFIGURATION: 'true'
+      AWS_ENDPOINT: 's3://s3:9000'
       PULSE_HOSTNAME: rabbitmq
       PULSE_USERNAME: admin
       PULSE_PASSWORD: admin
@@ -329,8 +348,10 @@ services:
       PULSE_HOSTNAME: rabbitmq
       PULSE_VHOST: local
       PULSE_AMQPS: 'false'
-      AWS_ACCESS_KEY_ID: ''
-      AWS_SECRET_ACCESS_KEY: ''
+      AWS_ACCESS_KEY_ID: 'minioadmin'
+      AWS_SECRET_ACCESS_KEY: 'miniopassword'
+      AWS_FORCE_PATH_STYLE: 'true'
+      AWS_SKIP_CORS_CONFIGURATION: 'true'
       ARTIFACT_REGION: local
       AWS_ENDPOINT: http://s3:9000
     command:
@@ -340,6 +361,8 @@ services:
         condition: service_healthy
       postgres:
         condition: service_healthy
+      s3_init_buckets:
+        condition: service_completed_successfully
     ports:
       - '3018:80'
   secrets-web:
```

With an image built from #5494 gets the queue service to start. Note that we might not want to delete buckets on every restart and instead use `--ignore-existing` on `mc mb`

**Comment by lotas** [comment]
Thanks for the help @Eijebong! I will apply your patch and test it

**Comment by Eijebong** [comment]
CSP got in the way for a while. My solution: remove it. Maybe we can do something better.

```

     location /api/auth {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://auth-web;
     }

     location /api/github {
-      proxy_pass http://github-web;
+      proxy_hide_header Content-Security-Policy;
+      proxy_pass http://$github_web;
     }

     location /api/hooks {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://hooks-web;
     }

     location /api/index {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://index-web;
     }

     location /api/notify {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://notify-web;
     }

     location /api/object {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://object-web;
     }

     location /api/purge-cache {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://purge-cache-web;
     }

     location /api/queue {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://queue-web;
     }

     location /api/secrets {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://secrets-web;
     }

     location /api/web-server {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://web-server-web;
     }

     location /api/worker-manager {
+      proxy_hide_header Content-Security-Policy;
       proxy_pass http://worker-manager-web;
     }
   }
```

I'm also using a trick forcing nginx to not resolve `github-web` on startup since that container doesn't start as it's missing config. Without that the ingress just falls on its face on startup.

I can now see the `Create Task` page. Fun stuff!

**Comment by lotas** [comment]
Indeed, CSP headers are baked-in in `nginx.conf` inside container. So it is either we overwrite those files (mounts) or just use docker-compose/nginx config to change those headers, which should be safe and easy for local dev.

**Comment by lotas** [comment]
Figured out `bad mac` is indeed mismatching hmac signatures.

`web-server-web` generates signuture using `host: ingress` while `auth-web` verifies signature using `host: auth-web`

todo: make them see the same host header

**Review by matt-boris — COMMENTED** [review]
Looks great! Such a great step for us!

Just left a few small comments.

File: infrastructure/tooling/src/generate/generators/docker-compose.js [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+      const procs = requirements[`procslist-${name}`];
+      // only web services for now
+
+      Object.keys(procs).forEach((proc) => {
+        if (procs[proc].type !== 'web') {
+          return;
+        }
+
+        const overrides = name === 'web-server' ? { ports: ['3050:3050'] } : {};
+        // hmm.. why does web-server ignore the PORT?
```
[the remark this answers is not in the document]
Comment (lotas) [reply infrastructure/tooling/src/generate/generators/docker-compose.js:316]: yep, thanks

File: infrastructure/tooling/src/generate/generators/docker-compose.js [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+  GITHUB_PRIVATE_PEM: '',
+  GITHUB_APP_ID: '',
+  WEBHOOK_SECRET: '',
+
+  // Notify
+  EMAIL_SOURCE_ADDRESS: 'root@local',
+
+  // Queue
+  PUBLIC_ARTIFACT_BUCKET: 'public-bucket', // minio?
+  PRIVATE_ARTIFACT_BUCKET: 'private-bucket', // minio?
```
[the remark this answers is not in the document]
Comment (lotas) [reply infrastructure/tooling/src/generate/generators/docker-compose.js:95]: leftovers

File: dev-docs/development-process.md [hunk 3]
```diff
@@ -199,3 +199,55 @@ It's rare to use this image during development, as running the image requires lo
...
+It is possible to run all web services locally using `docker-compose`.
+
+[`docker-compose.yml`](../docker-compose.yml) is being autogenerated with `yarn generate` (`yarn generate --target docker-compose.yml`)
+
+It defines dependencies: `postgres`, `rabbitmq`, `minio` (for S3 artifacts) and all web services.
+
+User names and passwords can be seen in [`docker-compose.yml`](../docker-compose.yml).
+
+```shell
+docker-compose up -d
```
[the remark this answers is not in the document]
Comment (lotas) [reply dev-docs/development-process.md:214]: good point

**Pull request merged**